src/esp/gfx/GenericDrawable.cpp

```diff
@@ -38,31 +38,55 @@ void GenericDrawable::setLightSetup(const Mn::ResourceKey& resourceKey) {
   updateShader();
 }
 
-void GenericDrawable::draw(const Mn::Matrix4& transformationMatrix,
-                           Mn::SceneGraph::Camera3D& camera) {
-  updateShader();
-
+void GenericDrawable::updateShaderLightingParameters(
+    const Mn::Matrix4& transformationMatrix,
+    Mn::SceneGraph::Camera3D& camera) {
   const Mn::Matrix4 cameraMatrix = camera.cameraMatrix();
 
-  std::vector<Mn::Vector3> lightPositions;
+  // todo: move to MaterialHelper or elsewhere
+  std::vector<Mn::Vector4> lightPositions;
   lightPositions.reserve(lightSetup_->size());
-  std::vector<Mn::Color4> lightColors;
+  std::vector<Mn::Color3> lightColors;
   lightColors.reserve(lightSetup_->size());
+  std::vector<Mn::Color3> lightSpecularColors;
+  lightSpecularColors.reserve(lightSetup_->size());
+  constexpr float dummyRange = 10000;
+  std::vector<float> lightRanges(lightSetup_->size(), dummyRange);
+  const Mn::Color4 ambientLightColor = getAmbientLightColor(*lightSetup_);
 
   for (Mn::UnsignedInt i = 0; i < lightSetup_->size(); ++i) {
-    lightPositions.emplace_back(getLightPositionRelativeToCamera(
-        (*lightSetup_)[i], transformationMatrix, cameraMatrix));
-
-    lightColors.emplace_back((*lightSetup_)[i].color);
+    const auto& lightInfo = (*lightSetup_)[i];
+    lightPositions.emplace_back(Mn::Vector4(getLightPositionRelativeToCamera(
+        lightInfo, transformationMatrix, cameraMatrix)));
+
+    const auto& lightColor = (*lightSetup_)[i].color;
+    lightColors.emplace_back(lightColor);
+
+    // In general, a light's specular color should match its base color.
+    // However, negative lights have zero (black) specular.
+    constexpr Mn::Color3 blackColor(0.0, 0.0, 0.0);
+    bool isNegativeLight = lightColor.x() < 0;
+    lightSpecularColors.emplace_back(isNegativeLight ? blackColor : lightColor);
   }
 
+  // See documentation in src/deps/magnum/src/Magnum/Shaders/Phong.h
   (*shader_)
-      .setAmbientColor(materialData_->ambientColor)
+      .setAmbientColor(materialData_->ambientColor * ambientLightColor)
```

Comment:
This is the object's ambientColor, which generally matches the PBR base color. And then I'm pre-multiplying it here with the ambient light color in the scene. @mosra let me know if this matches your expectation of what to pass to Phong::setAmbientColor . Note objects will often also have an ambient texture, which again generally matches the PBR base-color texture.

Reply:
Yes, that's how it's expected to be used. Please note the online docs have an example snippet as well, compared to the dry markup in the `Phong.h` file: https://doc.magnum.graphics/magnum/classMagnum_1_1Shaders_1_1Phong.html#Shaders-Phong-lights-ambient